java/velox4j/src/main/cpp/main/velox4j/memory/MemoryManager.cpp

```diff
@@ -0,0 +1,423 @@
+  }
+
+  uint64_t shrinkCapacity(
+      uint64_t targetBytes,
+      bool allowSpill,
+      bool allowAbort) override {
+    velox::memory::ScopedMemoryArbitrationContext ctx{};
+    velox::memory::MemoryReclaimer::Stats status;
+    velox::memory::MemoryPool* pool = nullptr;
+    {
+      std::unique_lock guard{mutex_};
+      VELOX_CHECK_EQ(
+          candidates_.size(),
+          1,
+          "ListenableArbitrator should only be used within a single root pool");
+      pool = candidates_.begin()->first;
+    }
+    pool->reclaim(
+        targetBytes,
+        memoryReclaimMaxWaitMs_,
+        status); // ignore the output
+    return shrinkCapacityInternal(pool, 0);
+  }
+
+  uint64_t shrinkCapacity(velox::memory::MemoryPool* pool, uint64_t targetBytes)
+      override {
+    return shrinkCapacityInternal(pool, targetBytes);
+  }
+
+  Stats stats() const override {
+    Stats stats; // no-op
+    return stats;
+  }
+
+  std::string toString() const override {
+    return fmt::format(
+        "ARBITRATOR[{}] CAPACITY {} {}",
+        kind_,
+        velox::succinctBytes(capacity()),
+        stats().toString());
+  }
+
+ private:
+  void growCapacityInternal(velox::memory::MemoryPool* pool, uint64_t bytes) {
+    // Since
+    // https://github.com/facebookincubator/velox/pull/9557/files#diff-436e44b7374032f8f5d7eb45869602add6f955162daa2798d01cc82f8725724dL812-L820,
+    // We should pass bytes as parameter "reservationBytes" when calling ::grow.
+    auto freeByes = pool->freeBytes();
+    if (freeByes > bytes) {
+      if (growPool(pool, 0, bytes)) {
+        return;
+      }
+    }
+    auto reclaimedFreeBytes = shrinkPool(pool, 0);
+    auto neededBytes = velox::bits::roundUp(
+        bytes - reclaimedFreeBytes, memoryPoolTransferCapacity_);
+    try {
+      listener_->allocationChanged(neededBytes);
+    } catch (const std::exception& e) {
+      // if allocationChanged failed, we need to free the reclaimed bytes
+      listener_->allocationChanged(-reclaimedFreeBytes);
+      throw;
+    }
+    auto ret = growPool(pool, reclaimedFreeBytes + neededBytes, bytes);
+    VELOX_CHECK(
+        ret,
+        "{} failed to grow {} bytes, current state {}",
+        pool->name(),
+        velox::succinctBytes(bytes),
+        pool->toString());
+  }
+
+  uint64_t shrinkCapacityInternal(
+      velox::memory::MemoryPool* pool,
+      uint64_t bytes) {
+    uint64_t freeBytes = shrinkPool(pool, bytes);
+    listener_->allocationChanged(-freeBytes);
+    return freeBytes;
+  }
+
+  const uint64_t memoryPoolInitialCapacity_; // FIXME: Unused.
+  const uint64_t memoryPoolTransferCapacity_;
+  const uint64_t memoryReclaimMaxWaitMs_;
+  AllocationListener* listener_{nullptr};
+
+  mutable std::mutex mutex_;
+  inline static std::string kind_ = "VELOX4J";
+  std::unordered_map<
+      velox::memory::MemoryPool*,
+      std::weak_ptr<velox::memory::MemoryPool>>
+      candidates_;
+};
+
+class ArbitratorFactoryRegister {
```

Comment:
Find the `ListenableArbitrator` and `ArbitratorFactoryRegister` are key to the memory manager. Would you document them?